substrate/client/service/src/client/notification_pinning.rs

```diff
@@ -0,0 +1,341 @@
+	}
+}
+
+impl<Block: BlockT, B: Backend<Block>> Limiter<Block::Hash, u32>
+	for UnpinningByLengthLimiter<Block, B>
+{
+	type KeyToInsert<'a> = Block::Hash;
+	type LinkType = usize;
+
+	fn is_over_the_limit(&self, length: usize) -> bool {
+		length > self.max_length
+	}
+
+	fn on_insert(
+		&mut self,
+		_length: usize,
+		key: Self::KeyToInsert<'_>,
+		value: u32,
+	) -> Option<(Block::Hash, u32)> {
+		log::debug!(target: LOG_TARGET, "Pinning block based on notification. hash = {key}");
+		if self.max_length > 0 {
+			Some((key, value))
+		} else {
+			None
+		}
+	}
+
+	fn on_replace(
+		&mut self,
+		_length: usize,
+		_old_key: &mut Block::Hash,
+		_new_key: Block::Hash,
+		_old_value: &mut u32,
+		_new_value: &mut u32,
+	) -> bool {
+		true
+	}
+
+	fn on_removed(&mut self, key: &mut Block::Hash, references: &mut u32) {
+		// If reference count was larger than 0 on removal,
+		// the item was removed due to capacity limitations.
+		// Since the cache should be large enough for pinned items,
+		// we want to know about these evictions.
+		if *references > 0 {
+			log::warn!(
+				target: LOG_TARGET,
+				"Notification block pinning limit reached. Unpinning block with hash = {key:?}"
+			);
+			if let Some(backend) = self.backend.upgrade() {
+				(0..*references).for_each(|_| backend.unpin_block(*key));
+			}
+		} else {
+			log::trace!(
+				target: LOG_TARGET,
+				"Unpinned block. hash = {key:?}",
+			)
+		}
+	}
+
+	fn on_cleared(&mut self) {}
+
+	fn on_grow(&mut self, _new_memory_usage: usize) -> bool {
+		true
+	}
+}
+
+/// Worker for the handling of notification pinning.
+///
+/// It receives messages from a receiver and pins/unping based on the incoming messages.
+/// All notification related unpinning should go through this worker. If the maximum number of
+/// notification pins is reached, the block from the oldest notification is unpinned.
+pub struct NotificationPinningWorker<Block: BlockT, Back: Backend<Block>> {
+	unpin_message_rx: TracingUnboundedReceiver<UnpinWorkerMessage<Block>>,
+	task_backend: Weak<Back>,
+	pinned_blocks: schnellru::LruMap<Block::Hash, u32, UnpinningByLengthLimiter<Block, Back>>,
+}
+
+impl<Block: BlockT, Back: Backend<Block>> NotificationPinningWorker<Block, Back> {
+	/// Creates a new `NotificationPinningWorker`.
+	pub fn new(
+		unpin_message_rx: TracingUnboundedReceiver<UnpinWorkerMessage<Block>>,
+		task_backend: Weak<Back>,
+	) -> Self {
+		let pinned_blocks =
+			schnellru::LruMap::<Block::Hash, u32, UnpinningByLengthLimiter<Block, Back>>::new(
+				UnpinningByLengthLimiter::new(NOTIFICATION_PINNING_LIMIT, task_backend.clone()),
+			);
+		Self { unpin_message_rx, task_backend, pinned_blocks }
+	}
+
+	pub fn handle_announce_message(&mut self, hash: Block::Hash) {
+		if let Some(entry) = self.pinned_blocks.get_or_insert(hash, Default::default) {
+			*entry = *entry + 1;
+		}
+	}
+
+	pub fn handle_unpin_message(&mut self, hash: Block::Hash) {
```

Comment:
```suggestion
	fn handle_announce_message(&mut self, hash: Block::Hash) {
		if let Some(entry) = self.pinned_blocks.get_or_insert(hash, Default::default) {
			*entry = *entry + 1;
		}
	}

	fn handle_unpin_message(&mut self, hash: Block::Hash) {
```

Don't need to be `pub`?